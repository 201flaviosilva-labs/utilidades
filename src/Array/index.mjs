import { allEqual } from "./allEqual.mjs";
import { choice } from "./choice.mjs";
import { findBigObject } from "./findBigObject.mjs";
import { findLowObject } from "./findLowObject.mjs";
import { isSorted } from "./isSorted.mjs";
import { moveLeft, moveRight } from "./Move.mjs";
import { shuffle } from "./shuffle.mjs";
import { sortAscending } from "./sortAscending.mjs";
import { sortAscendingObject } from "./sortAscendingObject.mjs";
import { sortDescending } from "./sortDescending.mjs";
import { sortDescendingObject } from "./sortDescendingObject.mjs";

/**
 * Functions utils for arrays
 * 
 * @namespace Array
 */
export {
	allEqual,
	choice,
	findBigObject,
	findLowObject,
	isSorted,
	moveLeft, moveRight,
	shuffle,
	sortAscending, sortAscendingObject,
	sortDescending, sortDescendingObject,
};
