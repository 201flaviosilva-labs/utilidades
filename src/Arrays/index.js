import { allEqual } from "./allEqual.js";
import { choice } from "./choice.js";
import { findBigObject } from "./findBigObject.js";
import { findLowObject } from "./findLowObject.js";
import { isSorted } from "./isSorted.js";
import { moveLeft, moveRight } from "./move.js";
import { shuffle } from "./shuffle.js";
import { sortAscending } from "./sortAscending.js";
import { sortAscendingObject } from "./sortAscendingObject.js";
import { sortDescending } from "./sortDescending.js";
import { sortDescendingObject } from "./sortDescendingObject.js";

/**
 * Functions utils for arrays
 * 
 * @namespace Arrays
 */
export {
	allEqual,
	choice,
	findBigObject,
	findLowObject,
	isSorted,
	moveLeft, moveRight,
	shuffle,
	sortAscending, sortAscendingObject,
	sortDescending, sortDescendingObject,
};
