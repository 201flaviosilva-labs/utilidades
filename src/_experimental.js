// Points
export function distanceTwoPoints(x1, y1, x2, y2) { return Math.sqrt((x1 - x2) * (x1 - x2) + (y1 - y2) * (y1 - y2)); }
export function distanceTwoPointsVector2(point1, point2) { return distanceTwoPoints(point1.x, point1.y, point2.x, point2.y); }

export function angleBetweenTwoPoints(x1, y1, x2, y2) { return Math.atan2(y2 - y1, x2 - x1); }
export function angleBetweenTwoPointsVector2(point1, point2) { return angleBetweenTwoPoints(point1.x, point1.y, point2.x, point2.y); }
export function angleBetweenTwoPointsDegrees(x1, y1, x2, y2) { return angleBetweenTwoPointsRadians(x1, y1, x2, y2) * 180 / Math.PI; }
export function angleBetweenTwoPointsVector2Degrees(point1, point2) { return angleBetweenTwoPointsDegrees(point1.x, point1.y, point2.x, point2.y); }

// Circles
export function distanceBetweenTwoCirclesFromCircle(circle1, circle2) {
	return distanceTwoPoints(circle1.x, circle1.y, circle2.x, circle2.y) - (circle1.radius + circle2.radius);
}
export function distanceBetweenTwoCircles(x1, y1, radius1, x2, y2, radius2) {
	return distanceBetweenTwoCirclesFromCircle(
		{ x: x1, y: y1, radius: radius1 },
		{ x: x2, y: y2, radius: radius2 }
	);
}

export function circleArea(radius) { return Math.PI * radius * radius; }
export function circlePerimeter(radius) { return 2 * Math.PI * radius; }

// Rectangles
export function distanceBetweenTwoRectanglesFromBounds(rect1, rect2) {
	const xOverlap = Math.min(rect1.x + rect1.width, rect2.x + rect2.width) - Math.max(rect1.x, rect2.x);
	const yOverlap = Math.min(rect1.y + rect1.height, rect2.y + rect2.height) - Math.max(rect1.y, rect2.y);
	return xOverlap * yOverlap;
}
export function distanceBetweenTwoRectangles(x1, y1, width1, height1, x2, y2, width2, height2) {
	return distanceBetweenTwoRectanglesFromBounds(
		{ x: x1 + width1 / 2, y: y1 + height1 / 2, width: width1, height: height1 },
		{ x: x2 + width2 / 2, y: y2 + height2 / 2, width: width2, height: height2 }
	);
}

export function angleBetweenTwoRectanglesFromBounds(rect1, rect2) {
	const xOverlap = Math.min(rect1.x + rect1.width, rect2.x + rect2.width) - Math.max(rect1.x, rect2.x);
	const yOverlap = Math.min(rect1.y + rect1.height, rect2.y + rect2.height) - Math.max(rect1.y, rect2.y);
	return Math.atan2(yOverlap, xOverlap);
}
export function angleBetweenTwoRectangles(x1, y1, width1, height1, x2, y2, width2, height2) {
	return angleBetweenTwoRectanglesFromBounds(
		{ x: x1 + width1 / 2, y: y1 + height1 / 2, width: width1, height: height1 },
		{ x: x2 + width2 / 2, y: y2 + height2 / 2, width: width2, height: height2 }
	);
}

export function rectangleArea(width, height) { return width * height; }
export function rectanglePerimeter(width, height) { return 2 * (width + height); }
export function rectanglePerimeterFromBounds(rectangle) { return rectanglePerimeter(rectangle.x, rectangle.y, rectangle.width, rectangle.height); }

export function rectangleVertices(x, y, width, height) {
	return [
		{ x: x, y: y },
		{ x: x + width, y: y },
		{ x: x + width, y: y + height },
		{ x: x, y: y + height }
	];
}
export function rectangleVerticesFromBounds(rectangle) { return rectangleVertices(rectangle.x, rectangle.y, rectangle.width, rectangle.height); }

export function rectangleCenterX(x, width) { return x + width / 2; }
export function rectangleCenterY(y, height) { return y + height / 2; }
export function rectangleCenter(x, y, width, height) { return { x: rectangleCenterX(x, width), y: rectangleCenterY(y, height) }; }
export function rectangleCenterXFromBounds(rectangle) { return rectangleCenterX(rectangle.x, rectangle.width); }
export function rectangleCenterYFromBounds(rectangle) { return rectangleCenterY(rectangle.y, rectangle.height); }
export function rectangleCenterFromBounds(rectangle) { return rectangleCenter(rectangle.x, rectangle.y, rectangle.width, rectangle.height); }
