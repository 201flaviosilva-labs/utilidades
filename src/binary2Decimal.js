/**
 * Convert a binary number to a decimal
 * 
 * @example
 * binary2Decimal("101010"); // 42
 * 
 * @param {string} binary - the value to be converted
 * @returns {number} - conversion decimal
 */
export function binary2Decimal(binary) {
	if (typeof binary === "string") return parseInt(binary, 2);
}
