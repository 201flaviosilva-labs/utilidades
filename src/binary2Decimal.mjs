/**
 * Convert a number/string to a decimal
 * 
 * @example
 * binary2Decimal(101010); // 42
 * 
 * @param {number|string} binary - the number/string to be converted
 * @returns {number} - conversion decimal
 */
export function binary2Decimal(binary) {
	if (typeof binary === "string") return parseInt(binary.split("").reverse().join(""), 2);
	else if (typeof binary === "number") return parseInt(binary, 2);
}
