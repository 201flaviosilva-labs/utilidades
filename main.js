import * as utilidades from "./src/main";

console.log(utilidades);

document.querySelector("#app").innerHTML = `
  <div>
    <a href="https://vitejs.dev" target="_blank">
      <img src="/vite.svg" class="logo" alt="Vite logo" />
    </a>
    <h1>Utilidades Testes</h1>
  </div>
`;
