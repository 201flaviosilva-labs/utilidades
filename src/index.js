import * as Arrays from "./Arrays/index.js";
import * as Maths from "./Maths/index.js";
import * as Physics from "./Physics/index.js";

import { allCharactersSame } from "./allCharactersSame.js";
import { binary2Decimal } from "./binary2Decimal.js";
import { BinarySearchTree, BinarySearchTreeInstance } from "./BinarySearchTree.js";
import { choiceTrend } from "./choiceTrend.js";
import { clone } from "./clone.js";
import { compare2Objects } from "./compare2Objects.js";
import { decimal2Binary } from "./decimal2Binary.js";
import { getDate } from "./getDate.js"; // getTime, getMilliseconds, getSeconds, getMinutes, getHours, getDay, getWeekDay, getWeek, getMonth, getYear, getDateFormatted 
import { EventSystem, EventSystemInstance } from "./EventSystem.js";
import { Fibonacci, fibonacciSequence, fibonacciUntil, fibonacciCustomSequence, recursiveFibonacci } from "./Fibonacci.js";
import { getVersion } from "./getVersion.js";
import { isObjectEmpty } from "./isObjectEmpty.js";
import { isFalsy, isTruthy } from "./isTruthyFalsy.js";
import { and, or, xor } from "./logicalOperators.js";
import { randomColor, randomColor0X, randomRGBColor, randomRGBAColor } from "./randomColor.js";
import { randomNumber, randomInt, randomFloat } from "./randomNumber.js";
import { randomString } from "./randomString.js";
import { randomWalk1D, randomWalk2D, randomWalk3D } from "./randomWalk.js";
import { reverseString } from "./reverseString.js";
import { Vector2 } from "./Vector2.js";

export {
	Arrays,
	Maths,
	Physics,

	BinarySearchTree, BinarySearchTreeInstance,
	EventSystem, EventSystemInstance,
	Fibonacci, fibonacciSequence, fibonacciUntil, fibonacciCustomSequence, recursiveFibonacci,
	Vector2,

	allCharactersSame,
	binary2Decimal,
	compare2Objects,
	decimal2Binary,
	choiceTrend,
	clone,
	getDate,
	getVersion,
	isObjectEmpty,
	isFalsy, isTruthy,
	and, or, xor, // logicalOperators
	randomColor, randomColor0X, randomRGBColor, randomRGBAColor,
	randomNumber, randomInt, randomFloat,
	randomString,
	randomWalk1D, randomWalk2D, randomWalk3D,
	reverseString,
};
