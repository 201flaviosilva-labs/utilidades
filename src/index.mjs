import * as Array from "./Array/index.mjs";
// import * as DOM from "./DOM/index.mjs";
import * as Maths from "./Maths/index.mjs";
// import * as SortingAlgorithms from "./SortingAlgorithms/index.mjs";

import { allCharactersSame } from "./allCharactersSame.mjs";
import { binary2Decimal } from "./binary2Decimal.mjs";
import { BinarySearchTree, BinarySearchTreeInstance } from "./BinarySearchTree.mjs";
import { stringToCamelCase, camelCaseToNormal } from "./camelCase.mjs";
import { stringToCapitalize, capitalizeToNormal } from "./capitalizeCase.mjs";
import { choiceTrend } from "./choiceTrend.mjs";
import { clone } from "./clone.mjs";
import { compare2Objects } from "./compare2Objects.mjs";
import { decimal2Binary } from "./decimal2Binary.mjs";
import { getDate } from "./getDate.mjs"; // getTime, getMilliseconds, getSeconds, getMinutes, getHours, getDay, getWeekDay, getWeek, getMonth, getYear, getDateFormatted 
import { EventSystem, EventSystemInstance } from "./EventSystem.mjs";
import { Fibonacci, fibonacciSequence, fibonacciUntil, fibonacciCustomSequence, recursiveFibonacci } from "./Fibonacci.mjs";
import { FIFO } from "./FIFO.mjs";
// import { getUrlParameter } from "./getUrlParameter.mjs";
// import { getVersion } from "./getVersion.mjs";
import { invertSentence, invertWords } from "./invertText.mjs";
import { isFalsy, isTruthy } from "./isTruthyFalsy.mjs";
import { stringToKebabCase, kebabCaseToNormal } from "./kebabCase.mjs";
import { and, or, xor } from "./logicalOperators.mjs";
import { LIFO } from "./LIFO.mjs";
import { randomColor, randomColor0X, randomRGBColor, randomRGBAColor } from "./randomColor.mjs";
import { randomNumber, randomInt, randomFloat } from "./randomNumber.mjs";
import { randomString } from "./randomString.mjs";
import { randomWalk1D, randomWalk2D, randomWalk3D } from "./randomWalk.mjs";
import { stringToScreamingSnakeCase, screamingSnakeCaseToNormal } from "./screamingSnakeCase.mjs";
import { stringToSnakeCase, snakeCaseToNormal } from "./snakeCase.mjs";
import { topDownCarMovimentation } from "./topDownCarMovimentation.mjs";
import { Vector2 } from "./Vector2.mjs";

export {
	Array,
	// DOM,
	Maths,
	// SortingAlgorithms,

	BinarySearchTree, BinarySearchTreeInstance,
	EventSystem, EventSystemInstance,
	Fibonacci, fibonacciSequence, fibonacciUntil, fibonacciCustomSequence, recursiveFibonacci,
	FIFO,
	LIFO,
	Vector2,

	allCharactersSame,
	binary2Decimal,
	stringToCamelCase, camelCaseToNormal,
	stringToCapitalize, capitalizeToNormal,
	choiceTrend,
	clone,
	compare2Objects,
	decimal2Binary,
	getDate,
	// getUrlParameter,
	// getVersion,
	invertSentence, invertWords,
	isFalsy, isTruthy,
	stringToKebabCase, kebabCaseToNormal,
	and, or, xor, // logicalOperators
	randomColor, randomColor0X, randomRGBColor, randomRGBAColor,
	randomNumber, randomInt, randomFloat,
	randomString,
	randomWalk1D, randomWalk2D, randomWalk3D,
	stringToScreamingSnakeCase, screamingSnakeCaseToNormal,
	stringToSnakeCase, snakeCaseToNormal,
	topDownCarMovimentation,
};
