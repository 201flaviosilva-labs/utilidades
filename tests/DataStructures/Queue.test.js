import { describe, expect, it } from "vitest";
import { DataStructures } from "./src/main";
const { Queue, QueueNode: Node } = DataStructures;

describe("DataStructures/Queue.js", () => {
	describe("Queue", () => {
		describe("constructor()", () => {
			it("creates a new Queue, starting with a value", () => {
				const s = new Queue(100);

				expect(s.first).toBeInstanceOf(Node);
				expect(s.first.value).toBe(100);
				expect(s.first.next).toBeNull();

				expect(s.last).toBeInstanceOf(Node);
				expect(s.last.value).toBe(100);
				expect(s.last.next).toBeNull();

				expect(s.size).toBe(1);

				expect(s.toArray()).toEqual([100]);
				expect(s).toMatchSnapshot();
			});

			it("creates a new Queue, starting with an array", () => {
				const s = new Queue([10, 20, 30]);

				expect(s.first.value).toBe(10);
				expect(s.last.value).toBe(30);
				expect(s.size).toBe(3);
				expect(s.toArray()).toEqual([10, 20, 30]);
				expect(s).toMatchSnapshot();
			});

			it("creates a new string Queue", () => {
				const s = new Queue(["A", "B", "C"]);

				expect(s.first.value).toBe("A");
				expect(s.last.value).toBe("C");
				expect(s.toArray()).toEqual(["A", "B", "C"]);
			});

			it("creates a new empty Queue", () => {
				const s = new Queue();

				expect(s.first).toBeNull();
				expect(s.last).toBeNull();
				expect(s.size).toBe(0);

				expect(s.toArray()).toEqual([]);
			});

			it("creates a new empty Queue, starting with an empty array", () => {
				const s = new Queue([]);

				expect(s.first).toBeNull();
				expect(s.last).toBeNull();
				expect(s.size).toBe(0);

				expect(s.toArray()).toEqual([]);
			});
		});

		describe("enqueue()", () => {
			it("adds a new value to the end of the queue", () => {
				const s = new Queue([10, 20, 30]);

				expect(s.enqueue(40)).toEqual(s);
				expect(s.last.value).toBe(40);
				expect(s.size).toBe(4);

				expect(s.toArray()).toEqual([10, 20, 30, 40]);
			});

			it("should add a value to an empty queue", () => {
				const s = new Queue();

				expect(s.enqueue(10)).toEqual(s);

				expect(s.first.value).toBe(10);
				expect(s.last.value).toBe(10);
				expect(s.size).toBe(1);

				expect(s.toArray()).toEqual([10]);
			});
		});

		describe("dequeue()", () => {
			it("removes and returns the first value in the queue", () => {
				const s = new Queue([10, 20, 30]);

				expect(s.dequeue()).toBe(10);
				expect(s.size).toBe(2);

				expect(s.toArray()).toEqual([20, 30]);
			});

			it("should return the first node in the queue", () => {
				const s = new Queue([10, 20, 30]);

				expect(s.dequeue(true)).toBeInstanceOf(Node);

				expect(s.toArray()).toEqual([20, 30]);
			});

			it("should remove the first value from a stack with only one value", () => {
				const s = new Queue(10);

				expect(s.dequeue()).toBe(10);
				expect(s.first).toBeNull();
				expect(s.last).toBeNull();
				expect(s.size).toBe(0);
				expect(s.toArray()).toEqual([]);
			});

			it("should return undefined if the queue is empty", () => {
				const s = new Queue();

				expect(s.dequeue()).toBeUndefined();
			});

			it("should remove all the values from the queue", () => {
				const s = new Queue([10, 20, 30]);

				expect(s.dequeue()).toBe(10);
				expect(s.dequeue()).toBe(20);
				expect(s.dequeue()).toBe(30);

				expect(s.first).toBeNull();
				expect(s.last).toBeNull();
				expect(s.size).toBe(0);
				expect(s.toArray()).toEqual([]);
			});
		});

		describe("clear()", () => {
			it("should clear the list", () => {
				const s = new Queue([10, 20, 30]);
				s.clear();
				expect(s.size).toBe(0);
				expect(s.first).toBeNull();
				expect(s.last).toBeNull();
				expect(s.toArray()).toEqual([]);
			});

			it("should clear an empty queue", () => {
				const s = new Queue();
				s.clear();
				expect(s.size).toBe(0);
				expect(s.first).toBeNull();
				expect(s.last).toBeNull();
				expect(s.toArray()).toEqual([]);
			});
		});

		describe("peek()", () => {
			it("should return the top value/node from the queue", () => {
				const s = new Queue([10, 20, 30]);
				expect(s.peek()).toBe(10);
				expect(s.peek(true).value).toBe(10);
				expect(s.peek(true)).toBeInstanceOf(Node);
			});

			it("should return undefined if the queue is empty", () => {
				const s = new Queue();
				expect(s.peek()).toBeUndefined();
			});
		});

		describe("isEmpty()", () => {
			it("should return true if the Queue is empty", () => {
				const s = new Queue();
				expect(s.isEmpty()).toBe(true);
			});

			it("should return false if the Queue is not empty", () => {
				const s = new Queue([10, 20, 30]);
				expect(s.isEmpty()).toBe(false);
			});
		});

		describe("toArray()", () => {
			it("should convert the Queue to an array", () => {
				const s = new Queue([10, 20, 30]);
				expect(s.toArray()).toEqual([10, 20, 30]);
			});

			it("should convert an empty Queue to an empty array", () => {
				const s = new Queue();
				expect(s.toArray()).toEqual([]);
			});
		})
	});

	describe("Node", () => {
		it("creates a new Node with a value", () => {
			const node = new Node(10);

			expect(node.value).toBe(10);
			expect(node.next).toBeNull();
			expect(node).toMatchSnapshot();
		});
	});
});
