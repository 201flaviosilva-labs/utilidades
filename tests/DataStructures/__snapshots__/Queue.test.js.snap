// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DataStructures/Queue.js > Node > creates a new Node with a value 1`] = `
Node {
  "next": null,
  "value": 10,
}
`;

exports[`DataStructures/Queue.js > Queue > constructor() > creates a new Queue, starting with a value 1`] = `
Queue {
  "first": Node {
    "next": null,
    "value": 100,
  },
  "last": Node {
    "next": null,
    "value": 100,
  },
  "size": 1,
}
`;

exports[`DataStructures/Queue.js > Queue > constructor() > creates a new Queue, starting with an array 1`] = `
Queue {
  "first": Node {
    "next": Node {
      "next": Node {
        "next": null,
        "value": 30,
      },
      "value": 20,
    },
    "value": 10,
  },
  "last": Node {
    "next": null,
    "value": 30,
  },
  "size": 3,
}
`;
